### ISSUE TITLE
Massive DHT traffic on adding big datastes

### ISSUE THREAD

**Issue by Kubuxu** [ipfs/go-ipfs#2828]
@Magik6k was trying to add cdnjs (something I did few months ago) he have hit https://github.com/ipfs/go-ipfs/issues/2823 but also reported that adding the dataset which was about 21GB created about 3TB of traffic.

@Magik6k could you give more precise data on this?

**Comment by whyrusleeping** [comment]
The solution i'm investigating here is batching together outgoing provide operations, combining findpeer queries and outgoing 'puts'. The issue is that it might require making some backwards incompatible changes to the dht protocol, which isnt really a huge problem since we have multistream

**Comment by magik6k** [comment]
* Network: 800 GiB Out, 2 TiB in
* While adding files it was using ~50/100Mbps
* Around 10-20 Files/s, depending on file size, usually around 10-100KB
* 2633788 files
* 2-3 days IIRC

**Comment by whyrusleeping** [comment]
@magik6k Thanks! I will keep you posted as we make improvements to this.

**Comment by parkan** [comment]
Similar report, while adding ~70,000 objects (~100k ea) we were maxing out our instance's traffic (70MBit in, 160MBit out) for over 14 hours (!!!), at which point someone killed the IPFS daemon.

Back of the envelope, this is at least 1TB outgoing for ~10GB files.

One particular problem: because the daemon was killed before the process completed, it seems that *almost none of the files were pinned*. Running `ipfs repo gc` deleted almost everything:

```
; before gc
NumObjects       2995522
RepoSize         20
RepoPath         /home/ubuntu/.ipfs
Version          fs-repo@4

; after gc
NumObjects       771
RepoSize         20
RepoPath         /home/ubuntu/.ipfs
Version          fs-repo@4
```

(side note, the chunk counts seem way too high too, almost 40x per image instead of expected 4x)

**Comment by whyrusleeping** [comment]
@parkan which version of ipfs are you using? This is definitely a known issue, but it has been improved slightly since 0.4.2

**Comment by parkan** [comment]
@whyrusleeping 0.4.3-rc3

**Comment by parkan** [comment]
@whyrusleeping my intuition is that there's some kind of context that's not being released before a queue is drained, which is supported by *none* of the files being pinned successfully after killing the client. Does this seem plausible?

**Comment by whyrusleeping** [comment]
@parkan are you killing the daemon before the add call is complete?

**Comment by parkan** [comment]
@whyrusleeping no, these are all individual add calls that complete reasonably quickly, load/network traffic happens w/o client interaction

**Comment by jbenet** [comment]
> One particular problem: because the daemon was killed before the process completed, it seems that almost none of the files were pinned. Running ipfs repo gc deleted almost everything:

That's a very serious bug to figure out. consistency is paramout here and that's much more important to get right.

> Similar report, while adding ~70,000 objects (~100k ea) we were maxing out our instance's traffic (70MBit in, 160MBit out) for over 14 hours (!!!), at which point someone killed the IPFS daemon.
>
> Back of the envelope, this is at least 1TB outgoing for ~10GB files.

This is caused by providing random access to all content, as discussed via email. for the large use case, we should look at recompiling go-ipfs _without_ creating a provider record for each object, and leverage direct connections. (FWIW, orbit on js-ipfs works without the DHT entirely, and i imagine your use case here could do very well if you make sure to connect the relevant nodes together so there's no discovery to happen. This is the way to do it before pubsub connects them for you).

**Comment by whyrusleeping** [comment]
You can check which objects are pinned after an add with: `ipfs pin ls --type=recursive`. 

It also seems really weird that your repo size is listed as `20` for both of those calls...

**Comment by whyrusleeping** [comment]
@parkan also, what is the process for adding files? is it just a single call to `ipfs add` ? or many smaller calls broken up? Are you passing any other flags?

**Comment by jbenet** [comment]
@parkan on chunk size, i'd be interested (separately) how well Rabin Fingerprint based chunking (`ipfs add -s <file>`) does for your images. If they're similar it should reduce the storage, but this is experimental and not proved to help. it's also slower on add, as it's got to do math when chunking.

**Comment by parkan** [comment]
@whyrusleeping ipfs pin ls was showing <500 objects pinned successfully, I don't think any of them are from this add attempt

also, small correction to my previous statement: not *all* `ipfs add` calls had completed, basically it was like this:

1) we add ~10k files with relatively fast return from ipfs add
2) adds slow down, taking >30s per file
3) daemon is killed

**Comment by parkan** [comment]
@whyrusleeping these are many separate adds, we basically write a temporary file (incidentally, being able to pin from stdin would be great), `ipfs add` it (no options) and wait for return

**Comment by whyrusleeping** [comment]
@parkan Ah, yeah. The pins arent written until the add is complete. An ipfs add call will only pin the root hash of the result of any given call to `ipfs add`.

**Comment by whyrusleeping** [comment]
> being able to pin from stdin would be great

```
➜  ~ echo QmdfTbBqBPQ7VNxZEYEj14VmRuZBkqFbiwReogJgS1zR1n | ipfs pin add
pinned QmdfTbBqBPQ7VNxZEYEj14VmRuZBkqFbiwReogJgS1zR1n recursively
```

**Comment by whyrusleeping** [comment]
When the adds slow down that far could you get me some of the performance debugging info from https://github.com/ipfs/go-ipfs/blob/master/docs/debug-guide.md ?

**Comment by parkan** [comment]
@whyrusleeping sorry, maybe I was being unclear: for 10k+ files the "add" part **was** completed (i.e. the call to `ipfs add` returned) but pin did not land.

re: stdin, I meant `ipfs add`, not `ipfs pin add` :)  Can I just pipe binary data into `ipfs add -`?

**Comment by whyrusleeping** [comment]
So, you did:
```
ipfs add -r <dir with 100k files in it>
...
...
...
added <some hash> <that dirs name>
```

Right?
Does that final hash from that add call show up in the output of `ipfs pin ls --type=recursive` ?

I'm also confused about what is meant by 'adds slow down' if the add call completed by that point. Was this a subsequent call to `ipfs add` with more files?


re: stdin: you can indeed just pipe binary data to ipfs add :)

```
➜  ~ echo "hello" | ipfs add
added QmZULkCELmmk5XNfCgTnCyFgAVxBRBXyDHGGMVoLFLiXEN QmZULkCELmmk5XNfCgTnCyFgAVxBRBXyDHGGMVoLFLiXEN
➜  ~ dd if=/dev/urandom bs=4M count=2 | ipfs add
 7.75 MB / ? [----------=---------------------------------------------------------] 2+0 records in
2+0 records out
8388608 bytes (8.4 MB, 8.0 MiB) copied, 0.571907 s, 14.7 MB/s
added QmbPmPSxWcQqJGcq7hHJ7CGMxCQ2zxpMpNHLxARePPNv6n QmbPmPSxWcQqJGcq7hHJ7CGMxCQ2zxpMpNHLxARePPNv6n
```

**Comment by parkan** [comment]
here's the parsed dump:

<details>
```
crypto/sha256.(*digest).Sum                                                                                         1
runtime.goexit                                                                                                      1
os/signal.signal_recv                                                                                               1
gx/ipfs/QmNQynaz7qfriSUJkiEZUrm2Wen1u3Kj9goZzWtrPyu7XR/go-log.(*MirrorWriter).logRoutine                            1
gx/ipfs/QmV3NSS3A1kX5s28r7yLczhDsXzkgo65cqRgKFXYunWZmD/metrics.init.1.func2                                         1
gx/ipfs/QmbBhyDKsY4mbY6xsKt3qu9Y7FPvMJ6qbD8AMjYYvPRw1g/goleveldb/leveldb/util.(*BufferPool).drain                   1
gx/ipfs/QmbBhyDKsY4mbY6xsKt3qu9Y7FPvMJ6qbD8AMjYYvPRw1g/goleveldb/leveldb.(*DB).mpoolDrain                           1
runtime.gopark                                                                                                      1
main.(*IntrHandler).Handle.func1                                                                                    1
gx/ipfs/QmeYJHEk8UjVVZ4XCRTZe6dFQrb8pGWD81LYCgeLp8CvMB/go-metrics.(*meterArbiter).tick                              1
gx/ipfs/QmduCCgTaLnxwwf9RFQy2PMUytrKcEH9msohtVxSBZUdgu/go-peerstream.(*Swarm).connGarbageCollect                    1
main.daemonFunc.func1                                                                                               1
gx/ipfs/QmQopLATEYMNg7dVqZRNDfeE2S1yKy8zrRh5xnYiuqeZBn/goprocess.(*process).CloseAfterChildren                      1
github.com/ipfs/go-ipfs/routing/dht/providers.(*ProviderManager).run                                                1
github.com/ipfs/go-ipfs/Godeps/_workspace/src/github.com/briantigerchow/pubsub.(*PubSub).start                      1
github.com/ipfs/go-ipfs/exchange/bitswap/decision.(*Engine).nextEnvelope                                            1
github.com/ipfs/go-ipfs/exchange/bitswap.(*WantManager).Run                                                         1
gx/ipfs/QmbBhyDKsY4mbY6xsKt3qu9Y7FPvMJ6qbD8AMjYYvPRw1g/goleveldb/leveldb.(*DB).jWriter                              1
github.com/ipfs/go-ipfs/exchange/bitswap.New.func2                                                                  1
github.com/ipfs/go-ipfs/mfs.(*Republisher).Run                                                                      1
github.com/ipfs/go-ipfs/exchange/bitswap.(*Bitswap).providerQueryManager                                            1
github.com/ipfs/go-ipfs/routing/dht.(*IpfsDHT).Bootstrap.func1                                                      1
github.com/ipfs/go-ipfs/exchange/bitswap.(*Bitswap).rebroadcastWorker                                               1
github.com/ipfs/go-ipfs/exchange/bitswap.(*Bitswap).provideCollector                                                1
github.com/ipfs/go-ipfs/exchange/bitswap.(*Bitswap).provideWorker                                                   1
github.com/ipfs/go-ipfs/namesys/republisher.(*Republisher).Run                                                      1
gx/ipfs/QmSscYPCcE1H3UQr2tnsJ2a9dK9LsHTBGgP71VW6fz67e5/mdns.(*client).query                                         1
gx/ipfs/QmbBhyDKsY4mbY6xsKt3qu9Y7FPvMJ6qbD8AMjYYvPRw1g/goleveldb/leveldb.(*DB).mCompaction                          1
gx/ipfs/QmbBhyDKsY4mbY6xsKt3qu9Y7FPvMJ6qbD8AMjYYvPRw1g/goleveldb/leveldb.(*DB).compactionError                      1
sync.(*Pool).Get                                                                                                    1
github.com/ipfs/go-ipfs/blocks/blockstore.(*blockstore).AllKeysChan.func2                                           1
gx/ipfs/QmTxLSvdhwg68WJimdS6icLPhZi28aTp6b7uihC2Yb47Xk/go-datastore/namespace.(*datastore).Query.func1              1
gx/ipfs/QmTxLSvdhwg68WJimdS6icLPhZi28aTp6b7uihC2Yb47Xk/go-datastore/keytransform.(*ktds).Query.func1                1
gx/ipfs/QmTxLSvdhwg68WJimdS6icLPhZi28aTp6b7uihC2Yb47Xk/go-datastore/query.ResultsWithChan.func1                     1
gx/ipfs/QmTxLSvdhwg68WJimdS6icLPhZi28aTp6b7uihC2Yb47Xk/go-datastore/flatfs.(*Datastore).Query.func1.1               1
gx/ipfs/QmbBhyDKsY4mbY6xsKt3qu9Y7FPvMJ6qbD8AMjYYvPRw1g/goleveldb/leveldb.(*DB).tCompaction                          1
github.com/ipfs/go-ipfs/commands.(*ChannelMarshaler).Read                                                           1
github.com/ipfs/go-ipfs/commands/http.internalHandler.ServeHTTP.func2                                               1
runtime/pprof.writeGoroutineStacks                                                                                  1
gx/ipfs/QmQdnfvZQuhdT93LNc5bos52wAmdr3G2p6G8teLJMEN32P/go-libp2p-peerstore.(*AddrManager).AddAddrs                  1
gx/ipfs/QmVCe3SNMjkcPgnpFhZs719dheq6xE7gJwjzV7aWcUM4Ms/go-libp2p/p2p/discovery.(*mdnsService).pollForEntries.func1  1
gx/ipfs/QmZ8MMKFwA95viWULoSYFZpA4kdFa8idmFSrP12YJwjjaL/yamux.(*Session).Ping                                        1
github.com/ipfs/go-ipfs/routing/dht.(*IpfsDHT).GetClosestPeers.func2                                                1
syscall.Syscall                                                                                                     1
main.daemonFunc                                                                                                     1
gx/ipfs/QmVCe3SNMjkcPgnpFhZs719dheq6xE7gJwjzV7aWcUM4Ms/go-libp2p/p2p/net/swarm.(*Swarm).addConnListener.func2       2
gx/ipfs/QmQopLATEYMNg7dVqZRNDfeE2S1yKy8zrRh5xnYiuqeZBn/goprocess/periodic.callOnTicker.func1                        2
gx/ipfs/QmVCe3SNMjkcPgnpFhZs719dheq6xE7gJwjzV7aWcUM4Ms/go-libp2p/p2p/net/conn.(*listener).Accept                    2
github.com/ipfs/go-ipfs/core/corehttp.Serve                                                                         2
main.merge.func1                                                                                                    2
gx/ipfs/QmdhsRK1EK2fvAz2i2SH5DEfkL6seDuyMYEsxKa9Braim3/client_golang/prometheus.computeApproximateRequestSize       2
gx/ipfs/QmduCCgTaLnxwwf9RFQy2PMUytrKcEH9msohtVxSBZUdgu/go-peerstream.(*Swarm).setupStream.func1                     4
gx/ipfs/Qmf91yhgRLo2dhhbc5zZ7TxjMaR1oxaWaoc9zRZdi1kU4a/go-multistream.(*lazyConn).readHandshake                     6
github.com/ipfs/go-ipfs/exchange/bitswap.(*Bitswap).taskWorker                                                      8
gx/ipfs/QmQopLATEYMNg7dVqZRNDfeE2S1yKy8zrRh5xnYiuqeZBn/goprocess.(*process).doClose.func1                           9
gx/ipfs/QmQopLATEYMNg7dVqZRNDfeE2S1yKy8zrRh5xnYiuqeZBn/goprocess.(*processLink).AddToChild                          10
gx/ipfs/QmZ8MMKFwA95viWULoSYFZpA4kdFa8idmFSrP12YJwjjaL/yamux.(*Session).waitForSendErr                              11
gx/ipfs/QmduCCgTaLnxwwf9RFQy2PMUytrKcEH9msohtVxSBZUdgu/go-peerstream.(*Swarm).removeStream.func1                    16
github.com/ipfs/go-ipfs/routing/dht.(*IpfsDHT).Provide.func1                                                        19
github.com/ipfs/go-ipfs/routing/dht.(*messageSender).ctxReadMsg                                                     25
gx/ipfs/QmX6DhWrpBB5NtadXmPSXYNdVvuLfJXoFNMvUMoVvP5UJa/go-context/io.(*ctxReader).Read                              27
gx/ipfs/QmQopLATEYMNg7dVqZRNDfeE2S1yKy8zrRh5xnYiuqeZBn/goprocess.(*process).Close                                   86
gx/ipfs/QmZ8MMKFwA95viWULoSYFZpA4kdFa8idmFSrP12YJwjjaL/yamux.(*Stream).Read                                         119
gx/ipfs/QmZ8MMKFwA95viWULoSYFZpA4kdFa8idmFSrP12YJwjjaL/yamux.(*Session).send                                        133
gx/ipfs/QmZ8MMKFwA95viWULoSYFZpA4kdFa8idmFSrP12YJwjjaL/yamux.(*Session).keepalive                                   143
github.com/ipfs/go-ipfs/exchange/bitswap.(*msgQueue).runQueue                                                       144
gx/ipfs/QmZ8MMKFwA95viWULoSYFZpA4kdFa8idmFSrP12YJwjjaL/yamux.(*Session).AcceptStream                                144
net.runtime_pollWait                                                                                                156
gx/ipfs/QmQopLATEYMNg7dVqZRNDfeE2S1yKy8zrRh5xnYiuqeZBn/goprocess.(*process).doClose                                 265
github.com/ipfs/go-ipfs/routing/dht.(*dhtQueryRunner).spawnWorkers                                                  425
gx/ipfs/QmQdnfvZQuhdT93LNc5bos52wAmdr3G2p6G8teLJMEN32P/go-libp2p-peerstore/queue.(*ChanQueue).process.func1         426
gx/ipfs/QmZy2y8t9zQH2a1b8q2ZSLKp17ATuJoCNxxyMFG5qFExpt/go-net/context.propagateCancel.func1                         426
gx/ipfs/QmQopLATEYMNg7dVqZRNDfeE2S1yKy8zrRh5xnYiuqeZBn/goprocess/context.CloseAfterContext.func1                    433
github.com/ipfs/go-ipfs/routing/dht.(*dhtQueryRunner).Run                                                           509
github.com/ipfs/go-ipfs/routing/dht.(*IpfsDHT).Provide                                                              511
sync.runtime_Semacquire                                                                                             2187
```
</details>

**Comment by parkan** [comment]
@whyrusleeping no, we are calling `ipfs add` *on each individual file* because there's no actual tree to add from

**Comment by whyrusleeping** [comment]
Thanks for the stack dump, that just confirms for me that the problems are caused by the overproviding. The 2000 gorotuines stuck in runtime_Semacquire are somewhat odd though, could I get the full stack dump to investigate that one further?

As for the adds, after each `ipfs add` call completes (or some subset of them complete) are those hashes visible in the output of `ipfs pin ls --type=recursive` ? *every* call to `ipfs add` should put an entry in there before it returns.

**Comment by parkan** [comment]
@whyrusleeping I'm trying to repro this behavior, but I am quite certain that we ended up with 10,000+ added but not pinned files in yesterday's run

Full stax: https://ipfs.io/ipfs/QmP344ugRiyZRKA2djLgE371MzW2FhwbghnJyH7hmbmAeQ

**Comment by whyrusleeping** [comment]
@parkan thank you! that stack dump is pretty gnarly, lots of lock contention trying to spew messages out to everyone on the dht, thats the providers problem alright. For that, as @jbenet says, we can make a separate build (for now, can merge in as an option later) that only provides the hashes you explicitly tell it to. 

As for the pinning issue, thats a very serious issue if what you say is correct. I will investigate and try to reproduce locally as well. cc @Kubuxu for an extra set of eyes

**Comment by parkan** [comment]
@whyrusleeping I naively tried to do `ipfs add --local $file`, but this still seems to provide; I am guessing `--local` really means "no network reads", not "no touching network", right?

**Comment by jbenet** [comment]
@parkan 
- are you doing `ipfs add` from the filesystem? possible for you to add offline? 
(ie `ipfs add` without daemon running, then turn it on when you're done)
- are you doing this in own network or public network? 
- possible for you do to it disconnected? not connected to the public dht while you do this huge add.
- if own network, the query should be exhausting quickly, odd if it's not.

**Comment by whyrusleeping** [comment]
Yeah, the `--local` flag doesn't work for all commands unfortunately. Its something we need to unify. It mostly means "don't read data from the network for this call"

**Comment by parkan** [comment]
@jbenet would love to have a chat about working around provider system ASAP. This dataset is <0.01% of our target scale so it doesn't sound like we can work with it.

**Comment by parkan** [comment]
I will try to repro the pinning failure case -- fairly confident it happened as I suspect, because I don't see any other way for all those objects to have ended up in the store but not pinned.

Separately, running `ipfs add ...` w/no daemon running seems to complete in reasonable and non-increasing time, is the data just coped into the store w/o announce in this mode?

**Comment by Kubuxu** [comment]
Yes, it will be re-announced periodically.

You can also run daemon in offline mode, `ipfs daemon --offline`.

**Comment by jbenet** [comment]
@parkan 

- yes, no announcements there until you turn on the daemon. When you turn it on, it will announce, but it's a periodic re-announce thing. It's still going to be a BW hog, but it will be less ridiculous. We still need to work around this.

- I strongly recommend doing this offline or in a separate, isolated network. Announcements there will be more reasonable and not bounded by the "dial latency" of connecting to very crap peers.

- yes, absolutely, let's work around this provider problem, we need to do it for everyone anyway, but the priority has been lower because other things have been requested more until now. Let's look deeper into this.

**Comment by parkan** [comment]
Ok I can confirm that the number of pinned files has fallen to 1 despite having 299392 objects in repo. Should I open a new issue for this?

**Comment by whyrusleeping** [comment]
@parkan yes please, and if you could provide repro steps

**Comment by jbenet** [comment]
@parkan could you make a git repo that reproduces this exact problem? would love to have it as a test case.

**Comment by parkan** [comment]
@jbenet it's a bit hard to repro but I'll do my best -- seems like a serious issue

**Comment by whyrusleeping** [comment]
@parkan does the daemon output any errors when this happens?

**Comment by whyrusleeping** [comment]
Alright, I made a few PRs to help out here with the providers problem, #3101 #3102 #3103 #3105 and #3106

**Comment by whyrusleeping** [comment]
A bit of an update here, I've gotten a few of these PRs merged, only one remaining. I ran some tests to check bandwidth usage here, good results so far: https://github.com/ipfs/go-ipfs/pull/3105#issuecomment-242183017

**Comment by parkan** [comment]
@whyrusleeping sorry missed your earlier question, no relevant errors that I could see. Improvements looking fantastic, though!

**Comment by whyrusleeping** [comment]
@parkan thanks! I've got a few more things up my sleeve that should help too, but theyre a bit farther out. In the meantime, you should now be able to do all your adds with the `--local` flag, and then manually do `ipfs dht provide` on the keys you want advertised to the dht. The provide command has a `-r` flag to recursively provide all child blocks too, but run without it will just provide the single given block.

Please let me know if anything weird comes up or you have any questions.

**Comment by parkan** [comment]
@whyrusleeping awesome! Great turnaround on this. Should we build from master to have these changes reflected?

**Comment by whyrusleeping** [comment]
@parkan yeah, you will have to build from master. I can also provide 'development' builds on the dists page if there is an interest in that

**Comment by parkan** [comment]
@whyrusleeping having unstable/dev/nightly binaries around would definitely be appreciated if it's not too much effort

**Comment by whyrusleeping** [comment]
@parkan sure thing! i'll push out some builds after #3105 merges, and start the discussion about automating some nightly releases

**Comment by parkan** [comment]
@whyrusleeping getting a nightly/CI build baked would be 👌

**Comment by jbenet** [comment]
Yeah, it would be great to have channels in distributions. filed it over at https://github.com/ipfs/distributions/issues/77

**Comment by hobofan** [comment]
@whyrusleeping I'd also really appreciate a development build, especially because of this issue. :)

**Comment by hobofan** [comment]
After adding 200MB of the Wikidata dataset with the current master version I have these stats:
```
Bandwidth
TotalIn: 11 GB
TotalOut: 2.0 GB
```

Compared to the previous report in https://github.com/ipfs/apps/issues/27#issuecomment-242562201 
thats an improvment of ~2.65x for inbound and ~3.2x for outbound traffic when using the current `master` compared to `0.4.2`.

The dataset consists of 24 million JSON files which are mostly 1-20KB big.

**Comment by casey** [comment]
Has the state of this issue changed since the last update?

I'm evaluating IPFS for a project, and if a 200MB dataset generates 13GB of traffic to add the data to IPFS, it will rule out IPFS as an option.

We would really like to be interoperable with IPFS and the broader IPFS ecosystem, but we'll have to use BitTorrent instead.

**Comment by whyrusleeping** [comment]
@casey you don't have to incur any traffic when adding data to ipfs. Adds can be done with the `--local` option which prevents any outbound DHT traffic during the add. 

Also, the outbound traffic usage without that should be far lower than when this issue was created.

cc @magik6k and @Stebalien for other input

**Comment by Stebalien** [comment]
> Also, the outbound traffic usage without that should be far lower than when this issue was created.

It should also get much lower in the upcoming release due to https://github.com/libp2p/go-libp2p-kad-dht/pull/182.

**Comment by casey** [comment]
@whyrusleeping Thanks for the response!

If I add data with `--local`, will other nodes on the network be able to find my node and download the data?

**Comment by Stebalien** [comment]
> If I add data with --local, will other nodes on the network be able to find my node and download the data?

Not unless they're already connected to your node, no.

**Comment by postables** [comment]
How do you invoke the `--local` option? I took a look at `ipfs add --help` and there was no `--local` option

**Comment by Stebalien** [comment]
It's a global flag, which, unfortunately, isn't added to the command-local help. You can find the help by running `ipfs --help`.

**Comment by casey** [comment]
Unfortunately this doesn't help our use case. Nodes must be able to discover one another in the DHT.

**Comment by postables** [comment]
@Stebalien  ah thank you!

@casey  you could look at using a [circuit relay](https://github.com/libp2p/specs/tree/master/relay) that might be able to solve the connectivity issue

**Comment by casey** [comment]
@postables Unfortunately, running relays between nodes is not feasible. We need connectivity between arbitrary nodes, and don't have the resources to expend on dedicated relays.

**Comment by casey** [comment]
As stated, relay nodes are not feasible. Without relay isn't feasible either, since all nodes are the same. This is a bittorrent-like use case, where all nodes are homogeneous peers, with no nodes having any particular role. As such, it's unclear how to determine which nodes would not use `--local`, and thus shoulder the burden of the excessive DHT traffic.

**Comment by postables** [comment]
Not sure how this is unfeasible, all you would need is a single node that runs with `--local`, and you upload files to that node.  You then directly connect your nodes that need the file to that `--local` node, and pull the data from there which should avoid the excessive DHT traffic.

  If none of these options are desirable for you, then IMO IPFS is not suitable for your use case at this time.

**Comment by casey** [comment]
I think that's probably about right. It's unfortunate, since we'd really like to be compatible with the IPFS ecosystem.

**Comment by postables** [comment]
IPFS is still fairly young, and didn't start gaining wide spread attraction until this year, so you might just have to wait a bit :+1:  IPFS  has been progressing pretty nicely lately so definitely keep an eye out.

**Comment by Stebalien** [comment]
Note: We *are* working on improving the provider (that's what this is called) logic as it obviously doesn't scale. However, it's on the back-burner at the moment while we finish up some of our in-progress endeavors.

**Comment by postables** [comment]
Do you have a link to the go code for the provider? Would be interesting to take a look at it.

**Comment by Stebalien** [comment]
It's kind of all over the place. This PR tries to improve that a bit: https://github.com/ipfs/go-ipfs/pull/4333

**Comment by casey** [comment]
Has there been any progress on this? Is this considered to be an issue, or is inserting/looking up each block of all file in the DHT going to be the behavior going forward?

**Comment by Stebalien** [comment]
This is considered an issue. The short term plan is to just announce fewer blocks (requires some changes to bitswap to avoid issues with restarting downlaods). The long-term plan is to introduce additional, more efficient content routing mechanisms.

**Comment by casey** [comment]
I think a simple fix would be to mirror BitTorrent, in the way that the hash of the info dict serves as the "topic" marker. BT clients look up peers via the infohash, and then expect those peers to have blocks. IPFS could do the same thing, use the hash of a "topic" block that contains the hashes, directly or transitively, and expect peers that they find with the topic block hash will have content blocks.

**Comment by Stebalien** [comment]
That's the "announce fewer blocks" plan. However, unlike bittorrent, there isn't really a true "root" hash. The best we can do is use the file root and the root of the request (e.g., use `QmFoo` `/ipfs/QmFoo/path/to/thing.txt`). Unfortunately:

1. Bitswap, the transfer protocol, doesn't know about the relationship between blocks, it just sees blocks. So, if I already _have_ `QmFoo`, but am missing a block inside `thing.txt`, it doesn't know to search for `QmFoo` again. This requires some modifications to pass this extra context to bitswap (planned).
2. Unfortunately, this kind of change will mean that two slightly different datasets will end up with entirely different "swarms". For example, if I publish wikipedia to IPFS, then update it, the root hashes will be different so peers looking for the new copy will never find peers with the old copy. However, 99% of the files are actually the same.

**Comment by casey** [comment]
My understanding of this is shaky, but if files are split into chunks automatically, then could the hash of the file serve as the topic hash for all the chunks? That would get get us down to one DHT lookup/insert per file.

**Comment by lidel** [comment]
I believe this was addressed in v0.39, please update to the [latest Kubo version](https://github.com/ipfs/kubo/releases/latest).

Since 0.39, the [ DHT Sweep provider](https://github.com/ipfs/kubo/blob/master/docs/config.md#providedhtsweepenabled) (now default via `Provide.DHT.SweepEnabled=true`) directly targets this problem.

### What shipped

- Sweep provider groups provides by keyspace region, batching DHT operations per peer instead of doing one-off provides for each block
- `ipfs add` no longer triggers immediate DHT provides for every block; blocks are queued for the sweep cycle
- Only root CIDs get fast-provided to the DHT immediately (`--fast-provide-root`, on by default)
- Configurable worker count (`Provide.DHT.MaxWorkers`) lets operators cap bandwidth usage

### Why it resolves this issue

- Original report: adding ~21GB created ~3TB of DHT traffic from per-block provide storms
- Sweep eliminates the traffic spike by distributing provides evenly over the reprovide interval
- Users who want even less traffic can set `Provide.Strategy=roots` to skip non-root blocks entirely

More details in release noted for 0.39 and at https://ipshipyard.com/blog/2025-dht-provide-sweep/
